File: apps/api/internal/votacao/sortear_test.go
package votacao_test

import (
	"errors"
	"math/rand"
	"testing"

	"github.com/PiluVitu/api/internal/votacao"
)

func sample() []votacao.SheetMovie {
	return []votacao.SheetMovie{
		{Number: 1, Title: "A Coisa", Type: "filme", Category: "terror", Watched: false},
		{Number: 2, Title: "Hereditário", Type: "filme", Category: "terror", Watched: false},
		{Number: 3, Title: "John Wick", Type: "filme", Category: "ação", Watched: true},
		{Number: 4, Title: "Breaking Bad", Type: "serie", Category: "drama", Watched: false},
		{Number: 5, Title: "Forrest Gump", Type: "filme", Category: "drama", Watched: false},
	}
}

func TestSortOnePerCategory_HappyPath(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{IncludeWatched: true}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 3 {
		t.Fatalf("got %d, want 3 categories", len(got))
	}
	seen := map[string]bool{}
	for _, m := range got {
		if seen[m.Category] {
			t.Errorf("duplicate category %q", m.Category)
		}
		seen[m.Category] = true
	}
}

func TestSortOnePerCategory_FilterByType(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{Types: []string{"serie"}}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 1 || got[0].Type != "serie" {
		t.Errorf("got %+v, want exactly the serie", got)
	}
}

func TestSortOnePerCategory_ExcludeWatched(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{IncludeWatched: false}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	for _, m := range got {
		if m.Watched {
			t.Errorf("watched movie leaked: %+v", m)
		}
	}
}

func TestSortOnePerCategory_IncludeWatched(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{IncludeWatched: true}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 3 {
		t.Errorf("got %d categories, want 3", len(got))
	}
}

func TestSortOnePerCategory_FilterByCategories(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{Categories: []string{"terror", "drama"}, IncludeWatched: true}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 2 {
		t.Fatalf("got %d, want 2", len(got))
	}
	for _, m := range got {
		if m.Category != "terror" && m.Category != "drama" {
			t.Errorf("unexpected category %q", m.Category)
		}
	}
}

func TestSortOnePerCategory_NoCandidates(t *testing.T) {
	_, err := votacao.SortOnePerCategory(nil, votacao.SortOptions{}, rand.New(rand.NewSource(1)))
	if !errors.Is(err, votacao.ErrNoCandidates) {
		t.Errorf("err = %v, want ErrNoCandidates", err)
	}
}

func TestSortOnePerCategory_NoCandidatesAfterFilter(t *testing.T) {
	movies := []votacao.SheetMovie{{Title: "Only", Type: "filme", Category: "terror", Watched: true}}
	_, err := votacao.SortOnePerCategory(movies, votacao.SortOptions{IncludeWatched: false}, rand.New(rand.NewSource(1)))
	if !errors.Is(err, votacao.ErrNoCandidates) {
		t.Errorf("err = %v, want ErrNoCandidates", err)
	}
}

func TestSortOnePerCategory_DeterministicWithSeed(t *testing.T) {
	movies := sample()
	a, _ := votacao.SortOnePerCategory(movies, votacao.SortOptions{IncludeWatched: true}, rand.New(rand.NewSource(42)))
	b, _ := votacao.SortOnePerCategory(movies, votacao.SortOptions{IncludeWatched: true}, rand.New(rand.NewSource(42)))
	if len(a) != len(b) {
		t.Fatal("length mismatch")
	}
	for i := range a {
		if a[i] != b[i] {
			t.Errorf("seed=42 not deterministic at index %d: %+v vs %+v", i, a[i], b[i])
		}
	}
}

func TestSortOnePerCategory_OrderedByCategoryAlpha(t *testing.T) {
	got, err := votacao.SortOnePerCategory(sample(), votacao.SortOptions{IncludeWatched: true}, rand.New(rand.NewSource(1)))
	if err != nil {
		t.Fatal(err)
	}
	wantCats := []string{"ação", "drama", "terror"}
	for i, w := range wantCats {
		if got[i].Category != w {
			t.Errorf("categories[%d] = %q, want %q", i, got[i].Category, w)
		}
	}
}

File: apps/ramielle/src/domain/sortear.test.ts
/**
 * Os 9 casos de `apps/api/internal/votacao/sortear_test.go` são o alvo desta
 * suíte (happy path, cada filtro isolado, sem candidatos ×2, determinismo,
 * ordenação) — mais testes específicos de TS pra provar que o `rng`
 * injetado (nunca `Math.random` direto) é o que decide a escolha dentro de
 * cada categoria.
 */
import { describe, expect, it, vi } from 'vitest'
import type { SheetMovie } from '../lib/gsheets'
import { NoCandidatesError, type Rng, sortOnePerCategory } from './sortear'

/** Espelha `sample()` de `sortear_test.go` — mesmos 5 filmes, mesma ordem. */
function sample(): SheetMovie[] {
  return [
    {
      number: 1,
      title: 'A Coisa',
      type: 'filme',
      category: 'terror',
      watched: false,
    },
    {
      number: 2,
      title: 'Hereditário',
      type: 'filme',
      category: 'terror',
      watched: false,
    },
    {
      number: 3,
      title: 'John Wick',
      type: 'filme',
      category: 'ação',
      watched: true,
    },
    {
      number: 4,
      title: 'Breaking Bad',
      type: 'serie',
      category: 'drama',
      watched: false,
    },
    {
      number: 5,
      title: 'Forrest Gump',
      type: 'filme',
      category: 'drama',
      watched: false,
    },
  ]
}

/**
 * PRNG determinístico (mulberry32) — só pra teste. Duas instâncias criadas
 * com a MESMA seed produzem a MESMA sequência de valores em `[0, 1)`; a
 * mesma instância NUNCA repete (é o que faz o teste de determinismo com
 * seed exigir DUAS instâncias, uma por chamada — mesma forma de
 * `rand.New(rand.NewSource(42))` ser criado duas vezes no Go).
 */
function seededRng(seed: number): Rng {
  let state = seed >>> 0
  return () => {
    state = (state + 0x6d2b79f5) >>> 0
    let t = state
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/** `rng` de fila — consome os valores passados, na ordem, e recicla. */
function queueRng(values: number[]): Rng {
  let i = 0
  return () => {
    const v = values[i % values.length]
    i += 1
    return v
  }
}

describe('sortOnePerCategory — happy path e filtros (paridade com sortear_test.go)', () => {
  it('happy path: 1 filme por categoria, sem duplicar categoria', () => {
    const got = sortOnePerCategory(
      sample(),
      { includeWatched: true },
      seededRng(1),
    )
    expect(got).toHaveLength(3)
    const categorias = new Set(got.map((m) => m.category))
    expect(categorias.size).toBe(3)
  })

  it('filtra por type — só sobra a série', () => {
    const got = sortOnePerCategory(sample(), { types: ['serie'] }, seededRng(1))
    expect(got).toHaveLength(1)
    expect(got[0]?.type).toBe('serie')
  })

  it('exclui assistidos por default (includeWatched ausente = false)', () => {
    const got = sortOnePerCategory(sample(), {}, seededRng(1))
    for (const m of got) {
      expect(m.watched).toBe(false)
    }
  })

  it('includeWatched:true inclui a categoria que só tem filme assistido', () => {
    const got = sortOnePerCategory(
      sample(),
      { includeWatched: true },
      seededRng(1),
    )
    expect(got).toHaveLength(3)
  })

  it('filtra por categories — só as categorias pedidas aparecem', () => {
    const got = sortOnePerCategory(
      sample(),
      { categories: ['terror', 'drama'], includeWatched: true },
      seededRng(1),
    )
    expect(got).toHaveLength(2)
    for (const m of got) {
      expect(['terror', 'drama']).toContain(m.category)
    }
  })

  it('lista vazia de entrada ⇒ NoCandidatesError', () => {
    expect(() => sortOnePerCategory([], {}, seededRng(1))).toThrow(
      NoCandidatesError,
    )
  })

  it('todos os filmes descartados pelo filtro ⇒ NoCandidatesError', () => {
    const movies: SheetMovie[] = [
      {
        number: 1,
        title: 'Only',
        type: 'filme',
        category: 'terror',
        watched: true,
      },
    ]
    expect(() =>
      sortOnePerCategory(movies, { includeWatched: false }, seededRng(1)),
    ).toThrow(NoCandidatesError)
  })

  it('determinístico com a mesma seed — duas chamadas independentes produzem a mesma saída', () => {
    const movies = sample()
    const a = sortOnePerCategory(
      movies,
      { includeWatched: true },
      seededRng(42),
    )
    const b = sortOnePerCategory(
      movies,
      { includeWatched: true },
      seededRng(42),
    )
    expect(a).toEqual(b)
  })

  it('categorias saem ordenadas alfabeticamente — ação, drama, terror', () => {
    const got = sortOnePerCategory(
      sample(),
      { includeWatched: true },
      seededRng(1),
    )
    expect(got.map((m) => m.category)).toEqual(['ação', 'drama', 'terror'])
  })
})

describe('sortOnePerCategory — o rng injetado decide a escolha (nunca Math.random)', () => {
  it('rng sempre 0 escolhe sempre o PRIMEIRO candidato de cada grupo, na ordem alfabética das categorias', () => {
    // Ordem de inserção no agrupamento (segue `sample()`): terror antes de
    // ação antes de drama. Ordenado por categoria: ação, drama, terror.
    const got = sortOnePerCategory(
      sample(),
      { includeWatched: true },
      queueRng([0, 0, 0]),
    )
    expect(got.map((m) => m.title)).toEqual([
      'John Wick', // ação — único candidato
      'Breaking Bad', // drama — primeiro da lista [Breaking Bad, Forrest Gump]
      'A Coisa', // terror — primeiro da lista [A Coisa, Hereditário]
    ])
  })

  it('rng perto de 1 escolhe sempre o ÚLTIMO candidato de cada grupo', () => {
    const got = sortOnePerCategory(
      sample(),
      { includeWatched: true },
      queueRng([0.999999, 0.999999, 0.999999]),
    )
    expect(got.map((m) => m.title)).toEqual([
      'John Wick', // ação — único candidato, índice 0 de qualquer forma
      'Forrest Gump', // drama — último da lista
      'Hereditário', // terror — último da lista
    ])
  })

  it('nunca chama Math.random — a escolha vem só do rng injetado', () => {
    const spy = vi.spyOn(Math, 'random')
    sortOnePerCategory(sample(), { includeWatched: true }, seededRng(7))
    expect(spy).not.toHaveBeenCalled()
    spy.mockRestore()
  })
})
